ulib/FStar.Matrix.fsti

```diff
@@ -0,0 +1,352 @@
+                   (eq:  CE.equiv c) ->
+                   (m: pos) -> (n: pos) ->
+                   CE.equiv (matrix c m n)                   
+                   
+val matrix_equiv_ijth (#c:_) (#m #n: pos) (eq: CE.equiv c) 
+                      (ma mb: matrix c m n) (i: under m) (j: under n)
+  : Lemma (requires (matrix_equiv eq m n).eq ma mb) 
+          (ensures ijth ma i j `eq.eq` ijth mb i j) 
+          
+val matrix_equiv_from_element_eq (#c:_) (#m #n: pos) (eq: CE.equiv c) (ma mb: matrix c m n)
+  : Lemma (requires (forall (i: under m) (j: under n). ijth ma i j `eq.eq` ijth mb i j))
+          (ensures matrix_eq_fun eq ma mb)
+ 
+let matrix_equiv_from_proof #c (#m #n: pos) (eq: CE.equiv c) (ma mb: matrix c m n)
+  (proof: (i:under m) -> (j:under n) -> Lemma (eq.eq (ijth ma i j) (ijth mb i j)))
+  : Lemma (matrix_eq_fun eq ma mb)
+  = Classical.forall_intro_2 proof; 
+    matrix_equiv_from_element_eq eq ma mb 
+
+let matrix_add_generator #c #eq (#m #n: pos) (add: CE.cm c eq) (ma mb: matrix c m n) 
+  : matrix_generator c m n = fun i j -> add.mult (ijth ma i j) (ijth mb i j)
+
+let matrix_add #c #eq (#m #n: pos) (add: CE.cm c eq) (ma mb: matrix c m n) 
+  : matrix_of (matrix_add_generator add ma mb)
+  = init (matrix_add_generator add ma mb)  
+  
+let matrix_add_ijth #c #eq (#m #n: pos) (add: CE.cm c eq) (ma mb: matrix c m n) (i: under m) (j: under n)
+  : Lemma (ijth (matrix_add add ma mb) i j == add.mult (ijth ma i j) (ijth mb i j)) = ()
+  
+val matrix_add_comm_monoid : (#c:Type) -> 
+                             (#eq:CE.equiv c) -> 
+                             (add: CE.cm c eq) -> 
+                             (m:pos) -> (n: pos) -> 
+                             CE.cm (matrix c m n) (matrix_equiv eq m n)
+
+let col #c #m #n (mx: matrix c m n) (j: under n) = SB.init m (fun (i: under m) -> ijth mx i j) 
+
+let row #c #m #n (mx: matrix c m n) (i: under m) = SB.init n (fun (j: under n) -> ijth mx i j) 
+
+val matrix_row_col_lemma (#c:_) (#m #n:pos) (mx: matrix c m n) (i: under m) (j: under n) 
+  : Lemma (ijth mx i j == SB.index (row mx i) j /\ ijth mx i j == SB.index (col mx j) i)  
+
+
+let seq_op_const #c #eq (cm: CE.cm c eq) (s: SB.seq c) (const: c) 
+  = SB.init (SB.length s) (fun (i: under (SB.length s)) -> cm.mult (SB.index s i) const)
+  
+let const_op_seq #c #eq (cm: CE.cm c eq) (const: c) (s: SB.seq c)                       
+  = SB.init (SB.length s) (fun (i: under (SB.length s)) -> cm.mult const (SB.index s i))
+
+let seq_of_products #c #eq (mul: CE.cm c eq) (s: SB.seq c) (t: SB.seq c {SB.length t == SB.length s})
+  = SB.init (SB.length s) (fun (i: under (SB.length s)) -> SB.index s i `mul.mult` SB.index t i)
+
+let seq_of_products_lemma #c #eq (mul: CE.cm c eq) (s: SB.seq c) (t: SB.seq c {SB.length t == SB.length s})
+  (r: SB.seq c{SB.equal r (SB.init (SB.length s) (fun (i: under (SB.length s)) -> SB.index s i `mul.mult` SB.index t i))})
+  : Lemma (seq_of_products mul s t == r) = ()
```

Comment:
proof can go to impl, even if it is only a ()

Reply:
Alright